src/geo/cartodb-layer-group-base.js

```diff
@@ -31,36 +31,51 @@ var CartoDBLayerGroupBase = Backbone.Model.extend({
     return false;
   },
 
+  /**
+   * Returns a TileJSON format object for the given layerIndex
+   * @param  {number} layerIndex The index of one of the CartoDB layers grouped by this class.
+   */
   getTileJSONFromTiles: function (layerIndex) {
     var urls = this.get('urls');
-    if (!urls) {
-      throw new Error('tileJSON for the layer cannot be calculated until urls are set');
+    if (urls) {
+      return {
+        tilejson: '2.0.0',
+        scheme: 'xyz',
+        grids: urls.grids[this._convertToMapnikLayerIndex(layerIndex)],
+        tiles: urls.tiles,
+        formatter: function (options, data) { return data; }
+      };
     }
+  },
 
-    return {
-      tilejson: '2.0.0',
-      scheme: 'xyz',
-      grids: urls.grids[this._getIndexOfVisibleLayer(layerIndex)],
```

Comment:
@xavijam: the problem was basically that this method was returning the wrong index. `urls.grids` is an array that contains one entry for each of the visible layers. In the case of Named Maps `this._getIndexOfVisibleLayer` was taking into account that Named Maps have a tiled layer at the "bottom" and the wrong grid was being included in the TileJSON. The new method `_convertToMapnikLayerIndex` only takes into account visible "mapnik" layers, which will (hopefully) match the index for the given `layerIndex` in `url.grids`.